### ISSUE TITLE
Scans in cache should drop dummy entries which fall into continuous ranges

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#8153]
Single-row queries for missing rows will populate cache with a dummy row entry to record the fact that the row is absent so that the next read for that key doesn't have to go to disk.

With time, a lot of such entries can accumulate in a partition. Scanning the partition may have to go through a lot of them to fill a page. This may badly impact latency of the read.

To minimize the effect of this, we should remove those dummy entries so that the next scan doesn't have to walk through them. Mind the MVCC rules.

**Comment by tgrabiec** [comment]
Point queries will actually leave not a dummy row, but a regular (but empty) row. That can cause a similar effect, except that the fill buffer will respect preemption. It still will take a while until we fill a query page with non-empty rows.

Dummy entries are inserted by scans which restrict the clustering key. A dummy row is inserted for each side of the clustering range, if it falls into discountinus (uncached) clustering range.

**Comment by slivne** [comment]
@tgrabiec do we have a counter for the dummy entries in the cache - so we can checkwe are running into this state ?

**Comment by tgrabiec** [comment]
@slivne No, we don't. I am about to send a patch which adds a counter for touched dummy entries.

**Comment by slivne** [comment]
@tgrabiec your patch went in - can we close this - whats left ?

**Comment by tgrabiec** [comment]
Zapping is done only when there is a single MVCC version of a partition in cache. This means that it will not be done (temporarily) during cache updates after memtable flush. If memtables are constantly flushing, the optimization will not be effective.

**Comment by tgrabiec** [comment]
The more general solution would be to zap also in `mutation_partition::apply_monotonically()`, which is invoked when MVCC versions are merged.

**Comment by tgrabiec** [comment]
Backported to 2020.1 and 2021.1
